src/tls/server.c

```diff
@@ -0,0 +1,902 @@
+ * @source: Output variable for DataSource; #CLIENT if fd is from
+ *          a client #Connection, or #WS if it came from a #WsInstance
+ */
+static Connection *
+get_fd_connection (int fd, enum DataSource* source)
+{
+  Connection* ret = NULL;
+  block_sigchld (true);
+  for (Connection *c = server.connections; c; c = c->next)
+    {
+      if (c->client_fd == fd)
+        {
+          *source = CLIENT;
+          ret = c;
+          break;
+        }
+      if  (c->ws_fd == fd)
+        {
+          *source = WS;
+          ret = c;
+          break;
+        }
+    }
+  block_sigchld (false);
+  return ret;
+}
+
+/**
+ * remove_connection: stop tracking and clean up connection(s)
+ *
+ * Remove all #Connections which either have the given @fd (client or ws), or
+ * the given #WsInstance. This happens when encountering EOF or SIGCHLD.
+ *
+ * @fd: file descriptor from client (browser) or ws connection, or ≤ 0 for "unspecified"
+ * @ws: If given, all connections related to this #WsInstance get removed
+ */
+static void
+remove_connection (int fd, WsInstance *ws)
+{
+  Connection *c, *cprev;
+  bool found = false;
+
+  block_sigchld (true);
+  for (c = server.connections, cprev = NULL; c; )
+    {
+      Connection *cnext = c->next;
+
+      if ( (fd > 0 && (c->client_fd == fd || c->ws_fd == fd)) || (ws && c->ws == ws) )
+        {
+          /* stop polling it */
+          if (epoll_ctl (server.epollfd, EPOLL_CTL_DEL, c->client_fd, NULL) < 0)
+            err (1, "Failed to remove epoll connection fd");
+          if (c->ws_fd)
+            {
+              if (epoll_ctl (server.epollfd, EPOLL_CTL_DEL, c->ws_fd, NULL) < 0)
+                err (1, "Failed to remove epoll connection ws fd");
+            }
+
+          /* remove connection from our list */
+          if (cprev == NULL) /* first connection */
+            server.connections = c->next;
+          else
+            cprev->next = c->next;
+
+          connection_free (c);
+          found = true;
+        }
+      else
+        cprev = c;
+
+      c = cnext;
+    }
+  block_sigchld (false);
+
+  if (!found)
+    debug ("remove_connection: fd %i or ws %s not found in connections", fd, ws ? ws->socket.sun_path : "(unset)");
+}
+
+/**
+ * connection_init_ws: Find or launch a cockpit-ws instance for a new Connection
+ *
+ * Find the responsible cockpit-ws instance for a new server connection, i. e.
+ * by connection type (https/http) and client-side certificate. If none exists,
+ * create one. Set server.wss or server.ws_notls.
+ */
+static void
+connection_init_ws (Connection *c)
+{
+  int fd;
+  const gnutls_datum_t *peer_der = NULL;
+  WsInstance *ws = NULL;
+  bool ws_add_tls = false;
+  bool ws_add_notls = false;
+  struct epoll_event ev = { .events = EPOLLIN };
+
+  if (c->is_tls)
+    {
+      peer_der = gnutls_certificate_get_peers (c->session, NULL);
+
+      /* find existing ws server for this peer cert */
+      block_sigchld (true);
```

Comment:
(As dicussed on IRC)

This will fork cockpit-ws with a blocked SIGCHLD, and sometimes this block makes it all the way to the cockpit-bridge processes, where it breaks things like cockpit.spawn that rely on getting SIGCHLD signals.

Reply:
In the integration tests, SIGCHLD doesn't seem to be blocked in the cockpit-bridge processes, but it happens in my devel VM, where everything runs as root.

Reply:
`setenforce 1` actually stops cockpit-bridge from inheriting the SIGCHLD blockage.